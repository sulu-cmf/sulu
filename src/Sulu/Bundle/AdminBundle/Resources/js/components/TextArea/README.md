A simple textarea component. The `onChange` callback is called whenever a change happens and the `onFinish` callback is
called as soon as the component loses its focus.

```javascript
initialState = {
    value: '',
};

const handleChange = (value) => {
    setState({
        value
    });
};

<TextArea
    value={state.value}
    placeholder="Tell me something about yourself..."
    onChange={handleChange}
    onFinish={() => alert('TextArea lost focus!')}
/>
```

It is also possible to pass a `maxCharacters` prop to show a [`CharacterCounter`](#charactercounter).

```javascript
initialState = {
    value: '',
};

const handleChange = (value) => {
    setState({
        value
    });
};

<TextArea
    maxCharacters={10}
    onChange={handleChange}
    value={state.value}
/>
```
