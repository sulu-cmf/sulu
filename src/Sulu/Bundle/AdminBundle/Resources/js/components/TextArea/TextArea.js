// @flow
import React, {Fragment} from 'react';
import classNames from 'classnames';
import CharacterCounter from '../CharacterCounter';
import textAreaStyles from './textArea.scss';

type Props = {|
    id?: string,
    maxCharacters?: number,
    name?: string,
    onBlur?: () => void,
    onChange: (string) => void,
    placeholder?: string,
    valid: boolean,
    disabled: boolean,
    value: ?string,
|};

export default class TextArea extends React.PureComponent<Props> {
    static defaultProps = {
        disabled: false,
        valid: true,
    };

    handleChange = (event: SyntheticEvent<HTMLInputElement>) => {
        this.props.onChange(event.currentTarget.value);
    };

    handleBlur = () => {
        const {onBlur} = this.props;

        if (onBlur) {
            onBlur();
        }
    };

    render() {
        const {
            id,
            disabled,
            maxCharacters,
            name,
            placeholder,
            value,
            valid,
        } = this.props;

        const textareaClass = classNames(
            textAreaStyles.textArea,
            {
                [textAreaStyles.error]: !valid,
                [textAreaStyles.disabled]: disabled,
            }
        );

        return (
            <Fragment>
                <textarea
                    className={textareaClass}
                    disabled={disabled}
                    id={id}
                    name={name}
                    onBlur={this.handleBlur}
                    onChange={this.handleChange}
                    placeholder={placeholder}
                    value={value || ''}
                />
                {maxCharacters &&
                    <CharacterCounter max={maxCharacters} value={value} />
                }
            </Fragment>
        );
    }
}
