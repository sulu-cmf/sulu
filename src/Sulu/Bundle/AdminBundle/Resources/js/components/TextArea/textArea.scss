@import '../../containers/Application/colors.scss';

$textareaDisabledColor: $gray;
$textareaDisabledBackgroundColor: $wildSand;
$textareaBorderColor: $silver;
$textareaErrorBorderColor: $persianRed;
$textareaBackgroundColor: $white;
$textareaColor: $black;

.text-area {
    width: 100%;
    min-height: 75px;
    background-color: $textareaBackgroundColor;
    color: $textareaColor;
    padding: 10px;
    border: 1px solid $textareaBorderColor;
    border-radius: 3px;
    resize: vertical;
    font-size: 12px;
    line-height: 20px;

    &.error {
        border-color: $textareaErrorBorderColor;
    }

    &.disabled {
        color: $textareaDisabledColor;
        background-color: $textareaDisabledBackgroundColor;
    }
}
