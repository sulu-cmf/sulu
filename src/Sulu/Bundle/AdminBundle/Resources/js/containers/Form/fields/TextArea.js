// @flow
import React from 'react';
import TextAreaComponent from '../../../components/TextArea';
import type {FieldTypeProps} from '../../../types';

export default class TextArea extends React.Component<FieldTypeProps<?string>> {
    render() {
        const {error, value, onChange, onFinish} = this.props;

        return (
            <TextAreaComponent
                onChange={onChange}
                onBlur={onFinish}
                value={value}
                valid={!error}
            />
        );
    }
}
