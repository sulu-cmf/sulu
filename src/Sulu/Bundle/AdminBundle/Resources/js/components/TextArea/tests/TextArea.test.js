// @flow
import React from 'react';
import {render, shallow} from 'enzyme';
import TextArea from '../TextArea';

jest.mock('../../../utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

test('TextArea should render', () => {
    expect(render(<TextArea onChange={jest.fn()} value="My value" />)).toMatchSnapshot();
});

test('TextArea should render with error', () => {
    expect(render(<TextArea onChange={jest.fn()} valid={false} value="My value" />))
        .toMatchSnapshot();
});

test('TextArea should render with placeholder', () => {
    expect(render(<TextArea onChange={jest.fn()} placeholder="My placeholder" value="My value" />))
        .toMatchSnapshot();
});

test('TextArea should render with value', () => {
    expect(render(<TextArea onChange={jest.fn()} value="My value" />)).toMatchSnapshot();
});

test('TextArea should render when disabled', () => {
    expect(render(<TextArea disabled={true} onChange={jest.fn()} value="My value" />)).toMatchSnapshot();
});

test('TextArea should render null value as empty string', () => {
    expect(render(<TextArea onChange={jest.fn()} value={null} />)).toMatchSnapshot();
});

test('TextArea should render with value and character counter', () => {
    expect(render(<TextArea maxCharacters={10} onChange={jest.fn()} value="My value" />)).toMatchSnapshot();
});

test('TextArea should call onBlur when it loses focus', () => {
    const blurSpy = jest.fn();
    const textArea = shallow(<TextArea onBlur={blurSpy} onChange={jest.fn()} value="" />);

    textArea.find('textarea').simulate('blur');
    expect(blurSpy).toBeCalledWith();
});

test('TextArea should call the callback when the TextArea changes', () => {
    const changeSpy = jest.fn();
    const textArea = shallow(<TextArea onChange={changeSpy} value="My value" />);
    textArea.find('textarea').simulate('change', {currentTarget: {value: 'my-value'}});
    expect(changeSpy).toHaveBeenCalledWith('my-value');
});
