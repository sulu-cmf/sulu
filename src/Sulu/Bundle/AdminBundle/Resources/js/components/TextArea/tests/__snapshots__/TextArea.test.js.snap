// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea should render 1`] = `
<textarea
  class="textArea"
>
  My value
</textarea>
`;

exports[`TextArea should render null value as empty string 1`] = `
<textarea
  class="textArea"
/>
`;

exports[`TextArea should render when disabled 1`] = `
<textarea
  class="textArea disabled"
  disabled=""
>
  My value
</textarea>
`;

exports[`TextArea should render with error 1`] = `
<textarea
  class="textArea error"
>
  My value
</textarea>
`;

exports[`TextArea should render with placeholder 1`] = `
<textarea
  class="textArea"
  placeholder="My placeholder"
>
  My value
</textarea>
`;

exports[`TextArea should render with value 1`] = `
<textarea
  class="textArea"
>
  My value
</textarea>
`;

exports[`TextArea should render with value and character counter 1`] = `
Array [
  <textarea
    class="textArea"
  >
    My value
  </textarea>,
  <label
    class="characterCounter"
  >
    2 sulu_admin.characters_left
  </label>,
]
`;
